engine/src/main/java/com/datatorrent/stram/engine/Node.java

```diff
@@ -485,6 +485,10 @@ protected void deactivateSinks()
 
   void checkpoint(long windowId)
   {
+    if (operator instanceof Operator.CheckpointNotificationListener) {
+      ((Operator.CheckpointNotificationListener)operator).beforeCheckpoint(windowId);
```

Comment:
Should this be caught for Throwable?
In case the beforeCheckpoint throws exception, the states which are meant to be saved won't be.